### PR TITLE
Support distribution metrics in Stackdriver

### PR DESCRIPTION
**Pull request by ZackButcher**
Adds support for distribution metrics in stackdriver, including client side aggregation of distribution metrics.

Note this PR is a superset of https://github.com/istio/mixer/pull/1235, so please review that one first.

**Release note**:

```release-note
NONE
```

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/1239)

### CODE REVIEW

File: adapter/stackdriver/config/config.proto [hunk 1]
```diff
@@ -62,6 +62,90 @@ message Params {
     message MetricInfo {
         google.api.MetricDescriptor.MetricKind kind = 1;
         google.api.MetricDescriptor.ValueType value = 2;
+
+        // Describes buckets for DISTRIBUTION valued metrics.
+        // TODO: look into using google.api.distribution.proto instead - we would pull it in but it has a ton of
+        // dependencies it doesn't actually use.
```
Comment (douglas-reid) [inline adapter/stackdriver/config/config.proto]: non-blocking: should we file an issue to track this?

File: adapter/stackdriver/helper/common.go [hunk 2]
```diff
@@ -35,3 +37,12 @@ func ToOpts(cfg *config.Params) (opts []gapiopts.ClientOption) {
 	}
 	return
 }
+
+// ToStringMap converts a map[string]interface{} to a map[string]string using fmt.Sprintf(%v)
+func ToStringMap(in map[string]interface{}) map[string]string {
```
Comment (douglas-reid) [inline adapter/stackdriver/helper/common.go]: curious: is this also in a current outstanding PR too?  I feel like i've seen this before, but I can't tell if i'm just going crazy.

File: adapter/stackdriver/log/log_test.go [hunk 3]
```diff
@@ -195,6 +196,18 @@ func TestHandleLogEntry(t *testing.T) {
...
 			}},
+		{"resource",
+			map[string]info{"resource": {tmpl: template.Must(template.New("").Parse("{{.a}}-{{.b}}-{{.c}}")), log: log}},
+			[]*logentry.Instance{{Name: "resource", Variables: map[string]interface{}{"a": 1, "b": "foo", "c": now}, MonitoredResourceType: "mr-type"}},
+			[]logging.Entry{
+				{
+					Timestamp: now,
+					Severity:  logging.Default,
+					Labels:    map[string]string{},
+					Payload:   fmt.Sprintf("%d-%s-%v", 1, "foo", now),
```
Comment (douglas-reid) [inline adapter/stackdriver/log/log_test.go]: super-minor-nit: any reason to not just hardcode '1' and `"foo"` in the `fmt` template ?

File: adapter/stackdriver/metric/merge.go [hunk 4]
```diff
@@ -107,19 +108,43 @@ func mergePoints(a, b *monitoring.TimeSeries) (*monitoring.TimeSeries, error) {
...
 	case *monitoring.TypedValue_DistributionValue:
-		if _, ok = b.Points[0].Value.Value.(*monitoring.TypedValue_DistributionValue); !ok {
+		var bv *monitoring.TypedValue_DistributionValue
+		if bv, ok = b.Points[0].Value.Value.(*monitoring.TypedValue_DistributionValue); !ok {
 			return a, fmt.Errorf("can't merge two timeseries with different value types; a is a distribution, b is not: %#v", b.Points[0].Value)
 		}
-		return a, fmt.Errorf("not implemented")
-		// TODO: combining distributions is hard. We know that they have the same buckets since they're all for the same metric, so that part is fine.
-		// But we need to update counts of each bucket, as well as the mean, sum of squared deviation, and the range. This is non-trivial.
+		// TODO: in theory we should assert that the DistributionValue's options are identical before merging, but given
```
Comment (douglas-reid) [inline adapter/stackdriver/metric/merge.go]: would this be worth logging as a warning, if detected ?

File: testdata/config/stackdriver.yaml [hunk 5]
```diff
@@ -55,34 +71,118 @@ spec:
...
+  monitoredResourceType: '"global"'
+  monitoredResourceDimensions:
+    # When using the global MR, this must be filled in, preferably matching the
+    # project_id given above in stackdriver config.
+    project_id: '"<your project ID>"'
+---
+apiVersion: "config.istio.io/v1alpha2"
+kind: metric
+metadata:
+  name: stackdriver-responsesize
```
Comment (douglas-reid) [inline testdata/config/stackdriver.yaml]: i missed the naming discussion resolution. Are we OK with dashed names ?  Should we update the other configs to have names like: `stackdriver-response-size` ?

File: adapter/stackdriver/config/config.proto (hunk 1 above)
Comment (ZackButcher) [reply adapter/stackdriver/config/config.proto]: https://github.com/istio/mixer/issues/1241

File: adapter/stackdriver/helper/common.go (hunk 2 above)
Comment (ZackButcher) [reply adapter/stackdriver/helper/common.go]: Yeah - this PR is a superset of https://github.com/istio/mixer/pull/1235, which is where this change is actually from.

File: adapter/stackdriver/metric/merge.go (hunk 4 above)
Comment (ZackButcher) [reply adapter/stackdriver/metric/merge.go]: I didn't want to bother checking it at runtime - we can add the check and return an error. I just don't really think it's necessary.

File: testdata/config/stackdriver.yaml (hunk 5 above)
Comment (ZackButcher) [reply testdata/config/stackdriver.yaml]: Yeah - dashes are fine, periods are not. I've verified this on multiple clusters. I think we should update the others too - @mandarjog thoughts?

Comment (ZackButcher) [reply testdata/config/stackdriver.yaml]: Talked with Kuat and removed all dashes. Apparently there's bugs in tooling that breaks dashes in some resource names.

File: adapter/stackdriver/log/log_test.go (hunk 3 above)
Comment (ZackButcher) [reply adapter/stackdriver/log/log_test.go]: Done

**Pull request merged**